stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+    /// state-machine to directly fetching it.  This only needs to happen if the tenure this state
+    /// machine is downloading contains the PoX anchor block, and it's also the last confirmed
+    /// tenurein this reward cycle.
+    ///
+    /// This function is called by `NakamotoTenureDownloadSet`, which instantiates, schedules, and
+    /// runs a set of these machines based on the peers' inventory vectors.  But because we don't
+    /// know if this is the PoX anchor block tenure (or even the last tenure) until we have
+    /// inventory vectors for this tenure's reward cycle, this state-transition must be driven
+    /// after this machine's instantiation.
+    pub fn transition_to_fetch_end_block(&mut self) -> Result<(), NetError> {
+        let NakamotoTenureDownloadState::WaitForTenureEndBlock(end_block_id, ..) = self.state
+        else {
+            return Err(NetError::InvalidState);
+        };
+        test_debug!(
+            "Transition downloader to {} to directly fetch tenure-end block {} (direct transition)",
+            &self.naddr,
+            &end_block_id
+        );
+        self.state = NakamotoTenureDownloadState::GetTenureEndBlock(end_block_id);
+        Ok(())
+    }
+
+    /// Transition to fetching the tenure-end block directly if waiting has taken too long.
+    pub fn try_transition_to_fetch_end_block(&mut self) {
+        if let NakamotoTenureDownloadState::WaitForTenureEndBlock(end_block_id, wait_deadline) =
+            self.state
+        {
+            if get_epoch_time_secs() < wait_deadline {
+                test_debug!(
+                    "Transition downloader to {} to directly fetch tenure-end block {} (timed out)",
+                    &self.naddr,
+                    &end_block_id
+                );
+                self.state = NakamotoTenureDownloadState::GetTenureEndBlock(end_block_id);
+            }
+        }
+    }
+
+    /// Validate and accept a tenure-end block.  If accepted, then advance the state.
+    /// Once accepted, this function extracts the tenure-change transaction and block header from
+    /// this block (it does not need the entire block).
+    ///
+    /// Returns Ok(()) if the block was valid
+    /// Returns Err(..) if the block was invalid
+    pub fn try_accept_tenure_end_block(
+        &mut self,
+        tenure_end_block: &NakamotoBlock,
+    ) -> Result<(), NetError> {
+        if !matches!(
+            &self.state,
+            NakamotoTenureDownloadState::WaitForTenureEndBlock(..)
+        ) && !matches!(
+            &self.state,
+            NakamotoTenureDownloadState::GetTenureEndBlock(_)
+        ) {
+            warn!("Invalid state for this method";
+                  "state" => %self.state);
+            return Err(NetError::InvalidState);
+        };
+        let Some(tenure_start_block) = self.tenure_start_block.as_ref() else {
+            return Err(NetError::InvalidState);
```

Comment:
Would you want to print a warning here as well? Seems a bit nebulous to have the same error type returned as InvalidState when the first error is checking state, but this error is checking if a tenure start block exists.

Reply:
Sure